radix-engine-common/src/math/decimal.rs

```diff
@@ -345,75 +345,166 @@ impl From<bool> for Decimal {
-where
-    <T as TryInto<Decimal>>::Error: fmt::Debug,
-{
+impl Sub<Decimal> for Decimal {
     type Output = Decimal;
 
-    fn sub(self, other: T) -> Self::Output {
+    #[inline]
+    fn sub(self, other: Decimal) -> Self::Output {
         let a = self.0;
-        let b_dec: Decimal = other.try_into().expect("Overflow");
-        let b: BnumI256 = b_dec.0;
-        let c: BnumI256 = a - b;
-        Decimal(c)
+        let b = other.0;
+        Decimal(a - b)
     }
 }
 
-impl<T: TryInto<Decimal>> Mul<T> for Decimal
-where
-    <T as TryInto<Decimal>>::Error: fmt::Debug,
-{
+impl Mul<Decimal> for Decimal {
     type Output = Decimal;
 
-    fn mul(self, other: T) -> Self::Output {
+    #[inline]
+    fn mul(self, other: Decimal) -> Self::Output {
         // Use BnumI384 (BInt<6>) to not overflow.
         let a = BnumI384::from(self.0);
-        let b_dec: Decimal = other.try_into().expect("Overflow");
-        let b = BnumI384::from(b_dec.0);
+        let b = BnumI384::from(other.0);
         let c = a * b / BnumI384::from(Self::ONE.0);
         let c_256 = BnumI256::try_from(c).expect("Overflow");
         Decimal(c_256)
     }
 }
 
-impl<T: TryInto<Decimal>> Div<T> for Decimal
-where
-    <T as TryInto<Decimal>>::Error: fmt::Debug,
-{
+impl Div<Decimal> for Decimal {
     type Output = Decimal;
 
-    fn div(self, other: T) -> Self::Output {
+    #[inline]
+    fn div(self, other: Decimal) -> Self::Output {
         // Use BnumI384 (BInt<6>) to not overflow.
         let a = BnumI384::from(self.0);
-        let b_dec: Decimal = other.try_into().expect("Overflow");
-        let b = BnumI384::from(b_dec.0);
+        let b = BnumI384::from(other.0);
         let c = a * BnumI384::from(Self::ONE.0) / b;
         let c_256 = BnumI256::try_from(c).expect("Overflow");
         Decimal(c_256)
     }
 }
 
-impl Neg for Decimal {
-    type Output = Decimal;
+macro_rules! impl_arith_ops {
+    ($type:ident) => {
+        impl Add<$type> for Decimal {
+            type Output = Decimal;
 
-    fn neg(self) -> Self::Output {
-        Decimal(-self.0)
+            fn add(self, other: $type) -> Self::Output {
+                self + Decimal::from(other)
+            }
+        }
+
+        impl Sub<$type> for Decimal {
+            type Output = Decimal;
+
+            fn sub(self, other: $type) -> Self::Output {
+                self - Decimal::from(other)
+            }
+        }
+
+        impl Mul<$type> for Decimal {
+            type Output = Decimal;
+
+            fn mul(self, other: $type) -> Self::Output {
+                self * Decimal::from(other)
+            }
+        }
+
+        impl Div<$type> for Decimal {
+            type Output = Decimal;
+
+            fn div(self, other: $type) -> Self::Output {
+                self / Decimal::from(other)
+            }
+        }
+
+        impl Add<Decimal> for $type {
+            type Output = Decimal;
+
+            fn add(self, other: Decimal) -> Self::Output {
```

Comment:
I believe these "flipped" methods would be smaller code size if they were all just `#[inline]` and the body was `other + self` - so it basically just reuses the other method, but with the arguments flipped.

Reply:
In case of:
a) multiplication addition flipping has no impact
b) substraction we would need to negate the second argument: `-other + self`
c) division we would have to invert the argument: `(Decimal::ONE /other) * self` and I am afraid there could appear some rounding issues, meaning we could receive different results than with current version `Decimal::from(self) / other`

So I would suggest to apply suggested change for a) and b)